kcidb/tools/kcidb_match.py

```diff
@@ -0,0 +1,471 @@
+
+        return {
+            'id': incident_id,
+            'origin': ORIGIN,
+            'issue_id': issue_id,
+            'issue_version': issue_version,
+            'present': True,
+            type_id_key: type_id_value,
+        }
+
+    def generate_incident_on_match(self, kcidb_io_object, issue_id,
+                                   issue_version, issue_pattern_object):
+        """Generate incident if issue pattern is found in a build/test
+        object"""
+        incidents = []
+
+        if self.validator.match_fields(issue_pattern_object, kcidb_io_object):
+            incident = self.create_incident(kcidb_io_object, issue_id,
+                                            issue_version)
+            incidents.append(incident)
+
+        return incidents
+
+    def generate_incidents_from_test(self, test):
+        """Generate incident from test object"""
+        kcidb_io_object = {"tests": [test._data],
+                           "builds": [test.build._data],
+                           "checkouts": [test.build.checkout._data]}
+        return self.generate_incidents_from_db(kcidb_io_object)
+
+    def generate_incidents_from_build(self, build):
+        """Generate incident from build object"""
+        kcidb_io_object = {"builds": [build._data],
+                           "checkouts": [build.checkout._data]}
+        return self.generate_incidents_from_db(kcidb_io_object)
+
+    def generate_incidents_from_db(self, kcidb_io_object):
+        incidents = []
+
+        for row in self.db.get_all_patterns():
+            issue_id, issue_version, pattern_object_json = row
+            pattern_object = json.loads(pattern_object_json)
+            incidents.extend(self.generate_incident_on_match(
+                kcidb_io_object, issue_id, issue_version, pattern_object))
+
+        return {
+            "version": KCIDB_IO_VERSION,
+            "incidents": incidents
+        }
+
+    def generate_incidents_from_object(self, kcidb_io_object):
+        incidents = []
+
+        for issue in kcidb_io_object['issues']:
+            pattern_object = issue['misc']['pattern_object'] \
+                if 'pattern_object' in issue['misc'] else {}
+            if not pattern_object:
+                continue  # Nothing to match
+            incidents.extend(self.generate_incident_on_match(
+                kcidb_io_object, issue['id'], issue['version'],
+                pattern_object))
+
+        return {
+            "version": KCIDB_IO_VERSION,
+            "incidents": incidents
+        }
+
+
+def fetch_kcidb_io_object(query_id, db_conn, query_type):
+    command = ["kcidb-query", f"-{query_type[0]}", query_id, "-d",
+               db_conn, "--parents"]
+    result = subprocess.run(command, capture_output=True, text=True)
+    if result.returncode != 0:
+        raise RuntimeError(f"Failed to fetch KCIDB IO object: {result.stderr}")
+    return json.loads(result.stdout)
+
+
+def validate_kcidb_io_object(kcidb_io_object):
```

Comment:
this validates if the given kcidb_io_object contains the minumum for this tool to work, but since we'll be receiving things from kcidb, we can drop this too

Reply:
The same comment as https://github.com/kernelci/kcidb/pull/586#discussion_r1787259604